### PR TITLE
[SPARK-30955][SQL] Exclude Generate output when aliasing in nested column pruning

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

When aliasing in nested column pruning in Project on top of Generate, we should exclude Generate outputs.

### Why are the changes needed?

Right now we would prune nested columns in Project on top of Generate. It is possible that referred nested columns are from Generate's outputs, not from its child. To address that case, we should exclude Generate outputs when aliasing in nested column pruning.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by viirya** [comment]
cc @dongjoon-hyun @peter-toth @maropu @HyukjinKwon @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala [hunk 1]
```diff
@@ -107,16 +107,17 @@ object NestedColumnAliasing {
    * 1. ExtractValue -> Alias: A new alias is created for each nested field.
    * 2. ExprId -> Seq[Alias]: A reference attribute has multiple aliases pointing it.
    */
-  def getAliasSubMap(exprList: Seq[Expression])
+  def getAliasSubMap(exprList: Seq[Expression], exclusiveAttrs: Seq[Attribute] = Seq.empty)
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala]: Ur, I see. This is a nit comment though, I feel sharing `getAliasSubMap` in both cases (`project w/generate case` and `simple project case`) makes the function arguments a bit complicated. So, its better to leave some comments about the argument `exclusiveAttrs` in the line 105.

**Comment by maropu** [comment]
I found a typo in the PR description: `should should` -> a single should.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 2]
```diff
@@ -3396,13 +3397,24 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
 
   test("SPARK-30870: Column pruning shouldn't alias a nested column if it means the whole " +
     "structure") {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: nit: this is not related to this pr though, for simplicity, can we pack the test title in a single line like `test("SPARK-30870: Column pruning shouldn't alias a nested column for the whole structure)`?

Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: ok.

Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala (hunk 1 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala]: Thanks. Added few comments.

**Comment by peter-toth** [comment]
Thanks @viirya for pinging me, the fix makes sense to me.

**Comment by viirya** [comment]
> I found a typo in the PR description: `should should` -> a single should.

Fixed. Thanks!

**Review by maropu — APPROVED** [review]
Looks fine cc: @dongjoon-hyun @dbtsai

**Comment by HyukjinKwon** [comment]
@viirya how does this relate to https://github.com/apache/spark/pull/27517?

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 3]
```diff
@@ -3394,15 +3395,25 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
...
-      """.stripMargin)
-    checkAnswer(df, Row(Row(1, 2)) :: Nil)
+  test("SPARK-30870: Column pruning shouldn't alias a nested column for the whole structure") {
+    withTable("t") {
+      val df = sql(
+        """
+          |SELECT value
+          |FROM VALUES array(named_struct('field', named_struct('a', 1, 'b', 2))) AS (value)
+        """.stripMargin)
+      df.write.format("parquet").saveAsTable("t")
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: The below test coverage looks better. Last time, it checked only the exception occurs or not. BTW, do we need to store as table to avoid some local table optimization?

**Comment by viirya** [comment]
> how does this relate to #27517?

This is to fix a bug when pruning nested column in Project on top of Generate.

#27517 extends the nested column pruning to a single Generate without Project.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 3 above)
Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Yea, if not store as a table, the query becomes a simple LocalRelation.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala [hunk 4]
```diff
@@ -104,19 +104,23 @@ object NestedColumnAliasing {
...
    * Return two maps in order to replace nested fields to aliases.
    *
+   * If `exclusiveAttrs` is given, any nested field accessors of these attributes
+   * won't be considered in nested fields aliasing.
+   *
    * 1. ExtractValue -> Alias: A new alias is created for each nested field.
    * 2. ExprId -> Seq[Alias]: A reference attribute has multiple aliases pointing it.
    */
-  def getAliasSubMap(exprList: Seq[Expression])
+  def getAliasSubMap(exprList: Seq[Expression], exclusiveAttrs: Seq[Attribute] = Seq.empty)
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala]: @viirya, just to clarify, we will keep `exclusiveAttrs` instead of `skipAttrs`?

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 3 above)
Comment (HyukjinKwon) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Sorry if I am being dumb here but I can't fully follow why LocalRelation matters here. How does the current fix relate to the test fix here?

If we concern about `ConvertToLocalRelation`, it seems not effective here:

```

== Parsed Logical Plan ==
'Project ['explodedvalue.field]
+- 'Generate 'explode('value), false, as, ['explodedvalue]
   +- 'SubqueryAlias AS
      +- 'UnresolvedInlineTable [value], [List('array('named_struct(field, 'named_struct(a, 1, b, 2))))]

== Analyzed Logical Plan ==
field: struct<a:int,b:int>
Project [explodedvalue#219.field AS field#220]
+- Generate explode(value#218), false, as, [explodedvalue#219]
   +- SubqueryAlias AS
      +- LocalRelation [value#218]

== Optimized Logical Plan ==
Project [explodedvalue#219.field AS field#220]
+- Generate explode(value#218), [0], false, as, [explodedvalue#219]
   +- LocalRelation [value#218]

== Physical Plan ==
*(1) Project [explodedvalue#219.field AS field#220]
+- Generate explode(value#218), false, [explodedvalue#219]
   +- LocalTableScan [value#218]
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala (hunk 4 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala]: I think yes, I will sync #27517 with this later.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 3 above)
Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: I changed the test added by SPARK-30870, because the root cause of the failure is not due to selecting whole structure. Here I change this test to more fit the original purpose of SPARK-30870: when selecting whole structure, column pruning shouldn't alias the nested column.

Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: The test added by SPARK-30870 is moved to `SchemaPruningSuite`.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala [hunk 5]
```diff
@@ -301,6 +301,24 @@ abstract class SchemaPruningSuite
     checkAnswer(query, Row("Y.", 1) :: Row("X.", 1) :: Row(null, 2) :: Row(null, 2) :: Nil)
   }
 
+  testSchemaPruning("select explode of nested field of array of struct from Generate output") {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala]: And, why is the test case here? Does it relate to source anything specific?

**Review by HyukjinKwon — APPROVED** [review]
The fix itself looks good if I am not mistaken.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 3 above)
Comment (peter-toth) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Both fixes fix the concrete SQL in SPARK-30870 independently because 
- the whole structure was selected (my pr fixes it) and 
- a nested column was selected from the output of a generate (this pr fixes it).

But this PR is better for cases that contain generate as it can handle nested field of a nested field https://github.com/apache/spark/pull/27702/files#diff-957112380b0a2ef014abc8227d0b70acR313-R318

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala (hunk 5 above)
Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala]: As the current source at `SchemaPruningSuite` doesn't have the data structure suitable for the test, this test doesn't use the source provided here. I put it here because it is highly related to nested column pruning. If we have another proper test suite to put, it is also ok.

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala]: Sorry, but I agree with @HyukjinKwon 's comment. This will be executed 16 times in `testSchemaPruning` x (ORCv1/ORCv2/Parquetv1/Parquetv2) without any benefit. If you need to add a E2E test, the original place is better than this.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Please move the test case back.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala (hunk 5 above)
Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala]: Ok. Let me move the test. Thanks.

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.0.

**Comment by dongjoon-hyun** [comment]
Thank you all.

**Comment by viirya** [comment]
Thanks!
